src/api/forms/service.test.js

```diff
@@ -873,10 +873,128 @@ describe('Forms service', () => {
         'Form is malformed in the database. Expected fields are missing.'
       )
     })
+
+    describe('with pagination', () => {
+      it('should return paginated results with correct pagination metadata', async () => {
+        const page = 1
+        const perPage = 2
+        const totalItems = 5
+
+        /** @type {WithId<Partial<FormMetadataDocument>>[]} */
+        const documents = [formMetadataFullDocument, formMetadataDraftDocument]
+
+        jest
+          .mocked(formMetadata.list)
+          .mockResolvedValue({ documents, totalItems })
+
+        const result = /** @type {Result<FormMetadata>} */ (
+          await listForms({ page, perPage })
+        )
+
+        expect(formMetadata.list).toHaveBeenCalledWith({ page, perPage })
+        expect(result).toEqual({
+          data: expect.any(Array),
+          meta: {
+            pagination: {
+              page,
+              perPage,
+              totalItems,
+              totalPages: Math.ceil(totalItems / perPage)
+            }
+          }
+        })
+        expect(result.data).toHaveLength(documents.length)
+      })
+
+      it('should return empty data when there are no forms', async () => {
+        const page = 1
+        const perPage = 10
+        const totalItems = 0
+
+        /** @type {WithId<Partial<FormMetadataDocument>>[]} */
+        const documents = []
+
+        jest
+          .mocked(formMetadata.list)
+          .mockResolvedValue({ documents, totalItems })
+
+        const result = /** @type {Result<FormMetadata>} */ (
+          await listForms({ page, perPage })
+        )
+
+        expect(formMetadata.list).toHaveBeenCalledWith({ page, perPage })
+        expect(result).toEqual({
+          data: [],
+          meta: {
+            pagination: {
+              page,
+              perPage,
+              totalItems,
+              totalPages: 0
+            }
+          }
+        })
+      })
+
+      it('should return correct totalPages when totalItems is not divisible by perPage', async () => {
+        const page = 1
+        const perPage = 3
+        const totalItems = 10
+
+        /** @type {WithId<Partial<FormMetadataDocument>>[]} */
+        const documents = [
+          formMetadataFullDocument,
+          formMetadataDraftDocument,
+          formMetadataLiveDocument
+        ]
+
+        jest
+          .mocked(formMetadata.list)
+          .mockResolvedValue({ documents, totalItems })
+
+        // Use a type assertion to inform the type checker
+        const result = /** @type {Result<FormMetadata>} */ (
```

Comment:
Ah has the `Promise<FormMetadata[] | Result<FormMetadata>>` return type made this tricky?

Wonder if it's [solveable with `@overload`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#overload-support-in-jsdoc) to set the correct return type based on params used?

Reply:
`@overload` actually works pretty well in this case 👌